vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/PriorityByWorkflow.java

```diff
@@ -0,0 +1,138 @@
+        return super.add(simpleEntry);
+      }
+
+      @Override
+      public boolean remove(Object o) {
+        try {
+          String oEntry = (String) o;
+          for (SimpleEntry entry : this){
+            if (entry.getKey().equals(oEntry)) {super.remove(entry);}
+          }
+        } catch (Exception e){
+          return false;
+        }
+        return true;
+      }
+
+      public SimpleEntry getByKey(String simpleEntryKey){
+        for (SimpleEntry entry : this){
+          if (entry.getKey().equals(simpleEntryKey)) {return(entry);}
+        }
+        return null;
+      }
+    };
+  }
+
+
+  //remove?
+  private static final Gauge currentInPriorityWaitingCount =
+      Gauge.build(
+              "vidarr_in_priority_waiting_per_workflow_current",
+              "The current number of workflows that are on priority waiting to run.")
+          .labelNames("workflow")
+          .register();
+
+  private final Map<String, WaitingState> workflows = new ConcurrentHashMap<>();
+
+  @Override
+  public Optional<Pair<String, BasicType>> inputFromSubmitter() {
+    return Optional.of(new Pair<String, BasicType>("priority", BasicType.INTEGER));
+  }
+
+  @Override
+  public void recover(String workflowName, String workflowVersion, String vidarrId, Optional<JsonNode> resourceJson) {
+    final var stateWaiting = workflows.computeIfAbsent(workflowName, k -> new WaitingState()).waiting;
+    // since we just created it if it doesn't exist, no need for null check here
+
+    int workflowPriority;
+
+    if (resourceJson.isPresent()) {
+      workflowPriority = resourceJson.get().asInt();
+    } else {
+      workflowPriority = 1;
+    }
+
+    stateWaiting.add(new SimpleEntry(vidarrId, workflowPriority));
+    currentInPriorityWaitingCount.labels(workflowName).set(stateWaiting.size());
+  }
+
+  @Override
+  public void release(String workflowName, String workflowVersion, String vidarrId) {
+    final var state = workflows.get(workflowName);
+    if (state != null) {
+
+      state.waiting.remove(vidarrId);
+      currentInPriorityWaitingCount.labels(workflowName).set(state.waiting.size());
+    }
+  }
+
+  @Override
+  public synchronized ConsumableResourceResponse request(
+      String workflowName, String workflowVersion, String vidarrId, Optional<JsonNode> input) {
+
+    int workflowPriority;
+    if (input.isEmpty()) {
+      workflowPriority = 1;
+    } else {
+      workflowPriority = input.get().asInt();
+    }
+
+    if (!acceptedPriorities.contains(workflowPriority)){
+      return ConsumableResourceResponse.error(
+          String.format("Vidarr error: The workflow %s run priority is invalid.", workflowName));
+    }
+    final var state = workflows.get(workflowName);
+    if (state == null) {
+      return ConsumableResourceResponse.error(
+          String.format("Internal Vidarr error: The %s workflow run priority has not been configured properly.", workflowName));
```

Comment:
I'm running into an issue where `state` is null for the first workflow run that is being checked. Do you have thoughts about what additional configuration I need to add?